core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -2759,11 +2761,20 @@ class ReplicaManager(val config: KafkaConfig,
                                           delta: TopicsDelta,
                                           topicId: Uuid): Option[(Partition, Boolean)] = {
     getPartition(tp) match {
-      case HostedPartition.Offline =>
-        stateChangeLogger.warn(s"Unable to bring up new local leader $tp " +
-          s"with topic id $topicId because it resides in an offline log " +
-          "directory.")
-        None
+      case HostedPartition.Offline(offlinePartition) =>
+        if (offlinePartition.flatMap(p => p.topicId).contains(topicId)) {
+          stateChangeLogger.warn(s"Unable to bring up new local leader $tp " +
+            s"with topic id $topicId because it resides in an offline log " +
+            "directory.")
+          None
+        } else {
+          stateChangeLogger.info(s"Creating new partition $tp with topic id " + s"$topicId." +
+            s"A topic with the same name but different id exists but it resides in an offline log " +
+            s"directory.")
```

Comment:
Question: From this check `if (offlinePartition.flatMap(p => p.topicId).contains(topicId))`, we can make sure this partition is not in an offline dir, but how could we know if the partition is an Online dir, or an None dir, or even in another offline dir? Should we use topicIDPartition as the key for `allPartitions`?

Reply:
I think the assumption is `allPartitions` only contains the TopicPartition key corresponding to the "latest" topic-id in case of a conflict.

I agree, keying it by `TopicIdPartition` would avoid ambiguity in the long run. Perhaps we should take that up in a future JIRA?

Reply:
What would be the impact if we don't fix it? 
Since currently, we're assuming the topicPartition with another topic ID is located on a dir with No state. So I'm thinking the impact would be that:
1. if the partition is on another dir which already has the Online state, we will create a duplicated one, or the state will be wrong.
2. if the partition is on another dir which is offline, we'll try to create it, but got exception without good handler (I think). 
3. Anything else?

Given this case is pretty rare that only happen when the topic name is the same, I agree we can create another JIRA for that improvement.

Reply:
Created https://issues.apache.org/jira/browse/KAFKA-16212